internal/encoding/merge/parquet.go

```diff
@@ -0,0 +1,330 @@
+		rows, err := blk.Select(blk.Schema())
+		if err != nil {
+			continue
+		}
+
+		// Fetch columns that is required by the static schema
+		cols := make(column.Columns, 16)
+		for name, typ := range schema {
+			col, ok := rows[name]
+			if !ok || col.Kind() != typ {
+				col = column.NewColumn(typ)
+			}
+
+			cols[name] = col
+		}
+
+		cols.FillNulls()
+
+		allCols := []column.Column{}
+		for _, colName := range schema.Columns() {
+			allCols = append(allCols, cols[colName])
+		}
+
+		for i := 0; i < allCols[0].Count(); i++ {
+			data := make(map[string]interface{})
+
+			j := 0
+			for _, colName := range schema.Columns() {
+				localCol := allCols[j]
+				fieldHandler := fieldHandlers[j]
+				finalData := localCol.At(i)
+
+				if finalData != nil && fieldHandler != nil {
+					finalData, _ = fieldHandler(localCol.At(i))
+				}
+
+				data[colName] = finalData
+
+				j++
+			}
+
+			if err := writer.AddData(data); err != nil {
+				return nil, errors.Internal("flush: error writing row", err)
+				// TODO: should we ignore or continue?
+			}
+		}
+	}
+
+	if err := writer.Close(); err != nil {
+		return nil, errors.Internal("flush: error closing writer", err)
+	}
+
+	// Always return a cloned buffer since we're reusing the working one
+	return clone(buffer), nil
+}
+
+type fieldHandler func(interface{}) (interface{}, error)
+
+func deriveSchema(inputSchema typeof.Schema) (schema *parquetschema.SchemaDefinition, fieldHandlers []fieldHandler, err error) {
+	schema = &parquetschema.SchemaDefinition{
+		RootColumn: &parquetschema.ColumnDefinition{
+			SchemaElement: &parquet.SchemaElement{
+				Name: "msg",
+			},
+		},
+	}
+
+	fieldHandlers = make([]fieldHandler, 0)
+
+	for _, field := range inputSchema.Columns() {
+		typ := inputSchema[field]
+
+		col, fieldHandler, err := createColumn(field, typ.String())
+		if err != nil {
+			return nil, nil, fmt.Errorf("couldn't create column for field %s: %v", field, err)
+		}
+
+		fieldHandlers = append(fieldHandlers, fieldHandler)
+		schema.RootColumn.Children = append(schema.RootColumn.Children, col)
+	}
+
+	if err := schema.Validate(); err != nil {
+		return schema, nil, fmt.Errorf("validation of generated schema failed: %w", err)
+	}
+
+	return schema, fieldHandlers, nil
+}
+
+func createColumn(field, typ string) (col *parquetschema.ColumnDefinition, fieldHandler func(interface{}) (interface{}, error), err error) {
+	col = &parquetschema.ColumnDefinition{
+		SchemaElement: &parquet.SchemaElement{},
+	}
+	col.SchemaElement.RepetitionType = parquet.FieldRepetitionTypePtr(parquet.FieldRepetitionType_OPTIONAL)
+	col.SchemaElement.Name = field
+
+	switch typ {
+	case "string":
+		col.SchemaElement.Type = parquet.TypePtr(parquet.Type_BYTE_ARRAY)
+		col.SchemaElement.LogicalType = parquet.NewLogicalType()
+		col.SchemaElement.LogicalType.STRING = &parquet.StringType{}
+		col.SchemaElement.ConvertedType = parquet.ConvertedTypePtr(parquet.ConvertedType_UTF8)
+		fieldHandler = byteArrayHandler
```

Comment:
For all of these, it's better to return right here to make it more readable `return optional(byteArrayHandler)`